### ISSUE TITLE
[Question] How to specify attributes to load on ProductRepository::getList()

### ISSUE THREAD

**Issue by Vinai** [magento/magento2#1341]
Using Magento version 0.74.0-beta12, I'm trying to load a list of products, some of which belong to the *Bag* attribute set of the sample data.

Roughly this is the code I'm playing with:
```php
        $this->searchCriteriaBuilder->addFilter($filters);
        $searchCriteria = $this->searchCriteriaBuilder->create();
        $productList = $this->productRepository->getList($searchCriteria);
```
As far as I can tell the `ProductRepository::getList()` method only adds the default attribute set attributes to the underlying collection.

```php
public function getList(\Magento\Framework\Api\SearchCriteriaInterface $searchCriteria)
{
    $collection = $this->collectionFactory->create();
    $defaultAttributeSetId = $this->eavConfig
        ->getEntityType(\Magento\Catalog\Api\Data\ProductAttributeInterface::ENTITY_TYPE_CODE)
        ->getDefaultAttributeSetId();
    $extendedSearchCriteria = $this->searchCriteriaBuilder->addFilter(
        [
            $this->filterBuilder
                ->setField('attribute_set_id')
                ->setValue($defaultAttributeSetId)
                ->create(),
        ]
    );
    foreach ($this->metadataService->getList($extendedSearchCriteria->create())->getItems() as $metadata) {
        $collection->addAttributeToSelect($metadata->getAttributeCode());
    }
```

How can I add additional attributes to the loaded products? As far as I can tell it is only possible to specify attribute filters using the `SearchCriteria`.
Or is it right to use `\Magento\Catalog\Model\Resource\Product\Collection` directly?

Thanks for clarifying.

**Comment by antonkril** [comment]
It's not right to use product collection directly, but this implementation of repository obviously does not support your scenario. So it should be fixed.

**Comment by Vinai** [comment]
Thanks for the reply Anton.
Would the following logic make sense in your opinion (pseudo code):
```
if (searchCriteriaHasAttributeSetFilter()) {
   attributeSetIds = getAttributeSetIdsFromSearchCriteria();
} else {
   attributeSetIds = [defaultAttributeSetId];
}
addAttributeSetAttributesToCollection(attributeSetIds);
```

**Comment by Vinai** [comment]
Is there an internal ticket for this issue?

**Comment by antonkril** [comment]
Yes. MAGETWO-38948. I don't think smth. is planned here. In my opinion attributeSetId and list of attributes to load do not belong to search criteria. These are operation modifiers and should be passed as an additional argument.

**Comment by Vinai** [comment]
From a clients point of view I would expect that each product received from the repository would have all values loaded for that product's attribute set, even if the `SearchResult` would contain products from different attribute sets.
This would also help decouple the internal storage logic further from any clients using the repository.
I'm aware implementing this without adding a couple more db roundtrips is not trivial.

**Comment by vpelipenko** [comment]
@Vinai, internal ticket is closed and fix is available in 1.0.0-beta. Please, check it on this version.

**Comment by Vinai** [comment]
From what I can see now all attributes that are assigned to at least one attribtue set are always added to be selected, This works, thanks!
